File: Data Structures & Algorithms/network-delay-time/submission-0.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<pair<int, int>>graph[101];

        for(auto &time: times){
            int u = time[0];
            int v = time[1];
            int dis = time[2];

            graph[u].push_back({v, dis});
        }
        queue<int>q;

        q.push(k);

        vector<int>distance(n+1, INT_MAX);
        distance[k] = 0;
        
        while(!q.empty()){
            int curr = q.front();
            q.pop();

            for(auto& neig : graph[curr]){
                int v = neig.first;
                int weight = neig.second;

                int currweight = distance[curr] + weight;

                if(distance[v] > currweight){
                    distance[v] = currweight;
                    q.push(v);
                }
            }
        }

        int ans = 0;

        for(int i=1;i<=n;i++){
            ans = max(ans, distance[i]);
        }

        return ans == INT_MAX? -1 : ans;
    }
};

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-0.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int[] dp = new int[n];
        if(n == 1)
            return 0;

        dp[0] = 0;
        dp[1] = Math.max(dp[0], prices[1]-prices[0]);

        for(int i=2;i<n;i++){
            int mx = 0;
            for(int j=i-1; j>=0; j--){
                int curr = prices[i] - prices[j] + (j <= 2 ? 0 : dp[j-2]);
                mx = Math.max(mx, curr);
            }
            dp[i] = Math.max(dp[i-1], mx);
        }

        return dp[n-1]; 
    }
}
